react/components/Session.tsx

```diff
@@ -1,63 +1,64 @@
-import React, { Component } from 'react'
+import React, { FunctionComponent, Fragment, useEffect, useRef, useState } from 'react'
 import Loading from './Loading'
-import { RenderContextProps, withRuntimeContext } from './RenderContext'
+import { useRuntime } from './RenderContext'
 
-interface State {
-  ensured: boolean
-  error: any
-}
-
-class Session extends Component<RenderContextProps, State> {
-  public state = { ensured: false, error: null }
-
-  public componentDidMount() {
-    this.onUpdate()
-  }
+function useSafeState(initialState: any) {
+  const [state, setState] = useState(initialState)
+  const mountedRef = useRef(false)
 
-  public componentDidUpdate() {
-    this.onUpdate()
-  }
-
-  public render() {
-    const { children } = this.props
-    const { ensured, error } = this.state
-
-    if (ensured) {
-      return children
+  useEffect(() => {
+    mountedRef.current = true
+    return () => {
+      mountedRef.current = false
     }
+  }, [])
 
-    if (error) {
-      return (
-        <div className="bg-washed-red pa6 f5 serious-black br3 pre">
-          <span>Error initializing session</span>
-          <pre>
-            <code className="f6">{error}</code>
-          </pre>
-        </div>
-      )
-    }
-
-    return <Loading />
+  const safeSetState = (value: any) => {
+    mountedRef.current && setState(value)
   }
 
-  private onUpdate() {
-    const {
-      runtime: { ensureSession },
-    } = this.props
-    const { ensured, error } = this.state
+  return [state, safeSetState]
+}
+
+const Session: FunctionComponent = ({ children }) => {
+  const [ensured, setEnsured] = useSafeState(false)
```

Comment:
I'd rather do something like [this](https://twitter.com/ryanflorence/status/1129487337375690752) instead of creating this custom hook, although I don't agree much with the naming :joy:

Reply:
LOVE IT